src/openfermion/utils/_trotter_qasm_conversion.py

```diff
@@ -0,0 +1,331 @@
+#	Licensed under the Apache License, Version 2.0 (the "License");
+#	you may not use this file except in compliance with the License.
+#	You may obtain a copy of the License at
+#
+#		http://www.apache.org/licenses/LICENSE-2.0
+#
+#	Unless required by applicable law or agreed to in writing, software
+#	distributed under the License is distributed on an "AS IS" BASIS,
+#	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#	See the License for the specific language governing permissions and
+#	limitations under the License.
+
+from openfermion.ops import QubitOperator
+import numpy as np
+
+'''
+Description:
+		Approximates the exponential of an operator composed of Pauli
+		matrices, by Trotterization. Outputs QASM format to a python
+		stream.
+
+		Change-of-basis gates:
+		H			: Z to X basis (and back)
+		Rx(pi/2)	: Z to Y basis ..... 
+		Rx(-pi/2) 	: Y to Z basis ..... 
+
+'''
+
+
+
+
+def third_order_trotter_helper(hamiltonian_dict,trotterized_list,op_a, op_b):
+	'''
+	This function recursively trotterizes a QubitOperator according
+		to the scheme:
+		e^(A+B) = e^(7/24 * A) e^(2/3 * B) e^(3/4 * A) e^(-1/24 * A) e^(B)
+
+	Used in trotterize() function.
+
+	Params: 
+		trotterized_list: list that the function populates with trotterized
+				QubitOperators
+		op_a: first term in hamiltonian as a QubitOperator
+		op_b: the rest of the terms in the hamiltonian as a list of
+				QubitOperator term dictionary keys
+
+	Returns: None
+	'''
+	trotterized_list.append(7.0/24.0 * op_a)
+	if len(op_b) == 1:
+		trotterized_list.append(2.0 / 3.0 * QubitOperator(op_b[0],
+			hamiltonian_dict[op_b[0]]))
+	else:
+		btemp = op_b[1:]
+		for i in btemp:
+			hamiltonian_dict[i] *= 2.0 / 3.0
+		atemp = QubitOperator(op_b[0],hamiltonian_dict[op_b[0]])
+		atemp *= 2.0 / 3.0
+		third_order_trotter_helper(trotterized_list, atemp, btemp)
+	trotterized_list.append(3.0/4.0 * op_a)
+
+	if op_b.terms.size == 1:
+		trotterized_list.append(-2.0/3.0 * QubitOperator(op_b[0],
+			hamiltonian_dict[op_b[0]]))
+	else:
+		btemp = op_b[1:]
+		for i in btemp:
+			hamiltonian_dict[i] *= -2.0 / 3.0
+		atemp = QubitOperator(op_b[0],hamiltonian_dict[op_b[0]])
+		atemp *= -2.0 / 3.0
+		third_order_trotter_helper(trotterized_list,atemp,btemp)
+	trotterized_list.append(-1.0 / 24.0 * op_a)
+
+	if op_b.terms.size == 1:
+		trotterized_list.append(QubitOperator(op_b[0],
+			hamiltonian_dict[op_b[0]]))
+	else:
+		btemp = op_b[1:]
+		atemp = QubitOperator(op_b[0],hamiltonian_dict[op_b[0]])
+		third_order_trotter_helper(trotterized_list, atemp, btemp)
+
+
+
+
+def trotterize(
+	hamiltonian,trotter_steps = 1, 
+	trotter_order = 1, 
+	term_ordering = [], 
```

Comment:
Do not initialize list parameters in functions in the arguments.  The behavior for this is well defined but very unexpected for most people, and leads to all kinds of usage errors and bugs.  Initialize to None and check if it has been passed with "is None" if you need this functionality.

Reply:
Looking through the code, I think this case actually would lead to a bug if we Trotterized several Hamiltonians in a row - Jarrod's roughly explained this, but if you're interested there are a bunch of articles online about this (mutable defaults in Python).

Reply:
Can you give me a case where it would lead to a bug (i.e. what you mean by several Hamiltonians in a row)? I'm going to change the function anyway but would still like to know when this one would fail.

Reply:
This is everybody's favorite unexpected python behavior. I was so confused when I first learned about this. There is this weird issue about passing mutable data structures as optional arguments. Here's an example: http://docs.python-guide.org/en/latest/writing/gotchas/.

Reply:
Ah I see... Wonder why python doesn't change that

Reply:
Still need to drop [] defaults?

Reply:
[]'s are dropped--that was still there because i didn't do "git rm" after changing the names of the files